File: frontend/lib/services/api/listing_service.dart
import 'package:easy_vacation/models/locations.model.dart';
import 'package:easy_vacation/models/stays.model.dart';
import 'package:easy_vacation/models/vehicles.model.dart';
import 'package:easy_vacation/models/activities.model.dart';
import 'package:easy_vacation/services/api/api_client.dart';
import 'package:easy_vacation/services/api/api_config.dart';
import 'package:easy_vacation/services/api/api_response.dart';

/// Listing model that combines Post with its category-specific details
class Listing {
  final int? id;
  final String ownerId; // Changed to String for UUID support
  final String category; // 'stay', 'vehicle', 'activity'
  final String title;
  final String? description;
  final double price;
  final String status;
  final String? availability;
  final Location location;
  final Stay? stayDetails;
  final Vehicle? vehicleDetails;
  final Activity? activityDetails;
  final List<String> images;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Listing({
    this.id,
    required this.ownerId,
    required this.category,
    required this.title,
    this.description,
    required this.price,
    this.status = 'draft',
    this.availability,
    required this.location,
    this.stayDetails,
    this.vehicleDetails,
    this.activityDetails,
    this.images = const [],
    this.createdAt,
    this.updatedAt,
  });

  factory Listing.fromJson(Map<String, dynamic> json) {
    try {
      print('🏠 Listing.fromJson parsing id=${json['id']}, images=${json['images']}');
      
      // Parse images - can be array of strings (URLs) or array of image objects with secure_url
      List<String> imageUrls = [];
      if (json['images'] != null) {
        final imagesData = json['images'];
        print('🏠 imagesData type: ${imagesData.runtimeType}, value: $imagesData');
        if (imagesData is List) {
          for (var img in imagesData) {
            print('🏠 Processing image: type=${img.runtimeType}, value=$img');
            if (img is String) {
              imageUrls.add(img);
            } else if (img is Map) {
              // Image object from post_images table - handle both Map<String, dynamic> and Map<dynamic, dynamic>
              final imgMap = Map<String, dynamic>.from(img);
              final url = imgMap['secure_url'] ?? imgMap['url'] ?? imgMap['content_url'];
              print('🏠 Extracted URL: $url');
              if (url != null) {
                imageUrls.add(url.toString());
              }
            }
          }
        }
      }
      print('🏠 Final imageUrls: $imageUrls');
      
      // Parse location - can be a Map or a List (from Supabase join)
      Map<String, dynamic> locationMap = {};
      if (json['location'] != null) {
        final locData = json['location'];
        if (locData is Map) {
          locationMap = Map<String, dynamic>.from(locData);
        } else if (locData is List && locData.isNotEmpty) {
          // Supabase sometimes returns joined data as array
          locationMap = Map<String, dynamic>.from(locData.first as Map);
        }
      }
      
      // Parse stay_details - can be a Map or a List
      Map<String, dynamic>? stayDetailsMap;
      if (json['stay_details'] != null) {
        final stayData = json['stay_details'];
        if (stayData is Map) {
          stayDetailsMap = Map<String, dynamic>.from(stayData);
        } else if (stayData is List && stayData.isNotEmpty) {
          stayDetailsMap = Map<String, dynamic>.from(stayData.first as Map);
        }
      }
      
      // Parse vehicle_details - can be a Map or a List
      Map<String, dynamic>? vehicleDetailsMap;
      if (json['vehicle_details'] != null) {
        final vehicleData = json['vehicle_details'];
        if (vehicleData is Map) {
          vehicleDetailsMap = Map<String, dynamic>.from(vehicleData);
        } else if (vehicleData is List && vehicleData.isNotEmpty) {
          vehicleDetailsMap = Map<String, dynamic>.from(vehicleData.first as Map);
        }
      }
      
      // Parse activity_details - can be a Map or a List
      Map<String, dynamic>? activityDetailsMap;
      if (json['activity_details'] != null) {
        final activityData = json['activity_details'];
        if (activityData is Map) {
          activityDetailsMap = Map<String, dynamic>.from(activityData);
        } else if (activityData is List && activityData.isNotEmpty) {
          activityDetailsMap = Map<String, dynamic>.from(activityData.first as Map);
        }
      }
      
      return Listing(
        id: json['id'],
        ownerId: json['owner_id']?.toString() ?? '',
        category: json['category'] ?? 'stay',
        title: json['title'] ?? '',
        description: json['description'],
        price: (json['price'] ?? 0).toDouble(),
        status: json['status'] ?? 'draft',
        availability: json['availability'],
        location: Location.fromMap(locationMap),
        stayDetails: stayDetailsMap != null 
            ? Stay.fromMap({...stayDetailsMap, 'post_id': json['id'] ?? 0})
            : null,
        vehicleDetails: vehicleDetailsMap != null 
            ? Vehicle.fromMap({...vehicleDetailsMap, 'post_id': json['id'] ?? 0})
            : null,
        activityDetails: activityDetailsMap != null 
            ? Activity.fromMap({...activityDetailsMap, 'post_id': json['id'] ?? 0})
            : null,
        images: imageUrls,
        createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : null,
        updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : null,
      );
    } catch (e) {
      print('❌ Error parsing Listing.fromJson: $e');
      print('📄 JSON data: $json');
      rethrow;
    }
  }

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'category': category,
      'title': title,
      'price': price,
      'location': location.toMap(),
    };

    if (id != null) map['id'] = id;
    if (description != null && description!.isNotEmpty) map['description'] = description;
    if (status.isNotEmpty) map['status'] = status;
    if (availability != null) map['availability'] = availability;
    if (images.isNotEmpty) map['images'] = images;

    // Add category-specific details
    if (stayDetails != null) {
      map['stay_details'] = {
        'stay_type': stayDetails!.stayType,
        'area': stayDetails!.area,
        'bedrooms': stayDetails!.bedrooms,
      };
    }

    if (vehicleDetails != null) {
      final vehicleMap = <String, dynamic>{
        'vehicle_type': vehicleDetails!.vehicleType,
      };
      // Only include optional fields if they have values
      if (vehicleDetails!.model != null && vehicleDetails!.model!.isNotEmpty) {
        vehicleMap['model'] = vehicleDetails!.model;
      }
      if (vehicleDetails!.year != null) vehicleMap['year'] = vehicleDetails!.year;
      if (vehicleDetails!.fuelType != null && vehicleDetails!.fuelType!.isNotEmpty) {
        vehicleMap['fuel_type'] = vehicleDetails!.fuelType;
      }
      if (vehicleDetails!.transmission != null) vehicleMap['transmission'] = vehicleDetails!.transmission;
      if (vehicleDetails!.seats != null) vehicleMap['seats'] = vehicleDetails!.seats;
      if (vehicleDetails!.features != null) vehicleMap['features'] = vehicleDetails!.features;
      map['vehicle_details'] = vehicleMap;
    }

    if (activityDetails != null) {
      map['activity_details'] = {
        'activity_type': activityDetails!.activityType,
        'requirements': activityDetails!.requirements,
      };
    }

    return map;
  }
}

/// Search filters for listings
class ListingFilters {
  final String? category;
  final String? wilaya;
  final String? city;
  final double? minPrice;
  final double? maxPrice;
  final String? stayType;
  final String? vehicleType;
  final String? activityType;
  final int limit;
  final int offset;

  ListingFilters({
    this.category,
    this.wilaya,
    this.city,
    this.minPrice,
    this.maxPrice,
    this.stayType,
    this.vehicleType,
    this.activityType,
    this.limit = 20,
    this.offset = 0,
  });

  Map<String, dynamic> toQueryParams() {
    final params = <String, dynamic>{};
    
    if (category != null) params['category'] = category;
    if (wilaya != null) params['wilaya'] = wilaya;
    if (city != null) params['city'] = city;
    if (minPrice != null) params['min_price'] = minPrice;
    if (maxPrice != null) params['max_price'] = maxPrice;
    if (stayType != null) params['stay_type'] = stayType;
    if (vehicleType != null) params['vehicle_type'] = vehicleType;
    if (activityType != null) params['activity_type'] = activityType;
    params['limit'] = limit;
    params['offset'] = offset;

    return params;
  }

  ListingFilters copyWith({
    String? category,
    String? wilaya,
    String? city,
    double? minPrice,
    double? maxPrice,
    String? stayType,
    String? vehicleType,
    String? activityType,
    int? limit,
    int? offset,
  }) {
    return ListingFilters(
      category: category ?? this.category,
      wilaya: wilaya ?? this.wilaya,
      city: city ?? this.city,
      minPrice: minPrice ?? this.minPrice,
      maxPrice: maxPrice ?? this.maxPrice,
      stayType: stayType ?? this.stayType,
      vehicleType: vehicleType ?? this.vehicleType,
      activityType: activityType ?? this.activityType,
      limit: limit ?? this.limit,
      offset: offset ?? this.offset,
    );
  }
}

/// Listing Service for managing listings with the backend
class ListingService {
  static ListingService? _instance;
  final ApiClient _apiClient;

  ListingService._internal() : _apiClient = ApiClient.instance;

  /// Get singleton instance
  static ListingService get instance {
    _instance ??= ListingService._internal();
    return _instance!;
  }

  /// Search/filter listings (public)
  /// 
  /// GET /api/listings
  Future<ApiResponse<PaginatedResponse<Listing>>> searchListings([ListingFilters? filters]) async {
    try {
      final response = await _apiClient.get(
        ApiConfig.listings,
        queryParams: filters?.toQueryParams(),
      );

      final listings = (response['data'] as List<dynamic>)
          .map((json) => Listing.fromJson(json))
          .toList();

      final meta = response['meta'] ?? {};
      
      return ApiResponse.success(PaginatedResponse(
        items: listings,
        total: meta['total'] ?? listings.length,
        limit: meta['limit'] ?? filters?.limit ?? 20,
        offset: meta['offset'] ?? filters?.offset ?? 0,
      ));
    } catch (e) {
      return ApiResponse.error(e.toString());
    }
  }

  /// Get a single listing by ID (public)
  /// 
  /// GET /api/listings/:id
  Future<ApiResponse<Listing>> getListingById(int id) async {
    try {
      final response = await _apiClient.get('${ApiConfig.listings}/$id');
      final listing = Listing.fromJson(response['data'] ?? response);
      return ApiResponse.success(listing);
    } catch (e) {
      return ApiResponse.error(e.toString());
    }
  }

  /// Get listings by owner ID (public)
  /// 
  /// GET /api/listings/owner/:ownerId
  Future<ApiResponse<List<Listing>>> getListingsByOwner(String ownerId) async {
    try {
      final response = await _apiClient.get('${ApiConfig.listings}/owner/$ownerId');
      final listings = (response['data'] as List<dynamic>)
          .map((json) => Listing.fromJson(json))
          .toList();
      return ApiResponse.success(listings);
    } catch (e) {
      return ApiResponse.error(e.toString());
    }
  }

  /// Get current user's listings (authenticated)
  /// 
  /// GET /api/listings/my/listings
  Future<ApiResponse<List<Listing>>> getMyListings() async {
    try {
      print('🔍 Calling /api/listings/my/listings...');
      final response = await _apiClient.get(
        '${ApiConfig.listings}/my/listings',
        requiresAuth: true,
      );
      
      print('📥 getMyListings response type: ${response.runtimeType}');
      print('📥 getMyListings response: $response');
      
      // Handle different response structures
      dynamic listingsData;
      if (response is Map && response.containsKey('data')) {
        print('📦 Response has "data" key');
        listingsData = response['data'];
        print('📦 Data type: ${listingsData.runtimeType}');
      } else if (response is List) {
        print('📦 Response is a List directly');
        listingsData = response;
      } else {
        print('⚠️ Unexpected response structure: ${response.runtimeType}');
        return ApiResponse.error('Unexpected response structure');
      }
      
      if (listingsData is! List) {
        print('⚠️ Data is not a list: ${listingsData.runtimeType}');
        return ApiResponse.error('Expected list of listings');
      }
      
      print('🔄 Parsing ${listingsData.length} listings...');
      final List<Listing> listings = [];
      for (int i = 0; i < listingsData.length; i++) {
        try {
          final json = listingsData[i];
          print('  🔸 Item $i type: ${json.runtimeType}');
          if (json is! Map) {
            print('  ⚠️ Item $i is not a Map, skipping');
            continue;
          }
          final listing = Listing.fromJson(Map<String, dynamic>.from(json));
          listings.add(listing);
          print('  ✅ Parsed listing: ${listing.title}');
        } catch (e) {
          print('  ❌ Error parsing item $i: $e');
        }
      }
      
      print('✅ Successfully parsed ${listings.length} listings');
      return ApiResponse.success(listings);
    } catch (e, stackTrace) {
      print('❌ getMyListings error: $e');
      print('❌ Stack trace: $stackTrace');
      return ApiResponse.error(e.toString());
    }
  }

  /// Create a new listing (authenticated)
  /// 
  /// POST /api/listings
  Future<ApiResponse<Listing>> createListing(Listing listing) async {
    try {
      print('🔑 Auth token present: ${_apiClient.isAuthenticated}');
      print('📤 Sending to: ${ApiConfig.listings}');
      
      final body = listing.toJson();
      print('📤 Body: $body');
      
      final response = await _apiClient.post(
        ApiConfig.listings,
        body: body,
        requiresAuth: true,
      );
      
      print('📥 Response: $response');
      final createdListing = Listing.fromJson(response['data'] ?? response);
      return ApiResponse.success(createdListing);
    } catch (e) {
      print('❌ createListing error: $e');
      // Try to extract more details from the error
      if (e.toString().contains('ValidationException')) {
        print('❌ Validation error details - check the body being sent');
      }
      return ApiResponse.error(e.toString());
    }
  }

  /// Update an existing listing (authenticated)
  /// 
  /// PATCH /api/listings/:id
  Future<ApiResponse<Listing>> updateListing(int id, Listing listing) async {
    try {
      final response = await _apiClient.patch(
        '${ApiConfig.listings}/$id',
        body: listing.toJson(),
        requiresAuth: true,
      );
      final updatedListing = Listing.fromJson(response['data'] ?? response);
      return ApiResponse.success(updatedListing);
    } catch (e) {
      return ApiResponse.error(e.toString());
    }
  }

  /// Delete a listing (authenticated)
  /// 
  /// DELETE /api/listings/:id
  Future<ApiResponse<void>> deleteListing(int id) async {
    try {
      await _apiClient.delete(
        '${ApiConfig.listings}/$id',
        requiresAuth: true,
      );
      return ApiResponse.success(null);
    } catch (e) {
      return ApiResponse.error(e.toString());
    }
  }

  /// Publish a listing (authenticated)
  /// 
  /// POST /api/listings/:id/publish
  Future<ApiResponse<Listing>> publishListing(int id) async {
    try {
      final response = await _apiClient.post(
        '${ApiConfig.listings}/$id/publish',
        requiresAuth: true,
      );
      final listing = Listing.fromJson(response['data'] ?? response);
      return ApiResponse.success(listing);
    } catch (e) {
      return ApiResponse.error(e.toString());
    }
  }

  /// Unpublish a listing (authenticated)
  /// 
  /// POST /api/listings/:id/unpublish
  Future<ApiResponse<Listing>> unpublishListing(int id) async {
    try {
      final response = await _apiClient.post(
        '${ApiConfig.listings}/$id/unpublish',
        requiresAuth: true,
      );
      final listing = Listing.fromJson(response['data'] ?? response);
      return ApiResponse.success(listing);
    } catch (e) {
      return ApiResponse.error(e.toString());
    }
  }

  /// Get listing availability (public)
  /// 
  /// GET /api/listings/:id/availability
  Future<ApiResponse<Map<String, dynamic>>> checkAvailability(int id, {
    DateTime? startDate,
    DateTime? endDate,
  }) async {
    try {
      final params = <String, dynamic>{};
      if (startDate != null) params['start_date'] = startDate.toIso8601String();
      if (endDate != null) params['end_date'] = endDate.toIso8601String();

      final response = await _apiClient.get(
        '${ApiConfig.listings}/$id/availability',
        queryParams: params,
      );
      return ApiResponse.success(response['data'] ?? response);
    } catch (e) {
      return ApiResponse.error(e.toString());
    }
  }

  /// Get unavailable dates for a listing (public)
  /// 
  /// GET /api/listings/:id/unavailable-dates
  Future<ApiResponse<List<DateTime>>> getUnavailableDates(int id) async {
    try {
      final response = await _apiClient.get('${ApiConfig.listings}/$id/unavailable-dates');
      final dates = (response['data'] as List<dynamic>?)
          ?.map((d) => DateTime.parse(d.toString()))
          .toList() ?? [];
      return ApiResponse.success(dates);
    } catch (e) {
      return ApiResponse.error(e.toString());
    }
  }

  /// Upload images to a listing (authenticated)
  /// 
  /// POST /api/listings/:id/images
  Future<ApiResponse<List<String>>> uploadImages(int id, List<String> filePaths) async {
    try {
      final images = <String>[];
      
      for (final path in filePaths) {
        final response = await _apiClient.uploadFile(
          '${ApiConfig.listings}/$id/images',
          filePath: path,
          fieldName: 'image',
          requiresAuth: true,
        );
        if (response['data'] != null && response['data']['url'] != null) {
          images.add(response['data']['url']);
        }
      }

      return ApiResponse.success(images);
    } catch (e) {
      return ApiResponse.error(e.toString());
    }
  }
}

File: backend/src/services/listing.service.ts
import supabase from '../config/supabase';
import { ApiError } from '../utils/apiError';
import { Post, Location, Stay, Vehicle, Activity, PostImage, Json } from '../types';
import { postsService } from './posts.service';
import { locationsService } from './locations.service';
import { staysService } from './stays.service';
import { vehiclesService } from './vehicles.service';
import { activitiesService } from './activities.service';
import { imageService } from './image.service';

export interface CreateListingInput {
  // Post fields
  owner_id: string;
  category: 'stay' | 'vehicle' | 'activity';
  title: string;
  description?: string;
  price: number;
  status?: string;
  availability?: string;

  // Location
  location: {
    wilaya: string;
    city: string;
    address: string;
    latitude: number;
    longitude: number;
  };

  // Category-specific details
  stay_details?: {
    stay_type: string;
    area?: number;
    bedrooms?: number;
  };

  vehicle_details?: {
    vehicle_type: string;
    model?: string;
    year?: number;
    fuel_type?: string;
    transmission?: boolean;
    seats?: number;
    features?: unknown;
  };

  activity_details?: {
    activity_type: string;
    requirements?: unknown;
  };

  // Images as base64 strings
  images?: string[];
}

export interface ListingWithDetails extends Post {
  location: Location;
  images: PostImage[];
  stay_details?: Stay;
  vehicle_details?: Vehicle;
  activity_details?: Activity;
  average_rating?: number;
  review_count?: number;
}

export interface SearchFilters {
  category?: string;
  wilaya?: string;
  city?: string;
  min_price?: number;
  max_price?: number;
  stay_type?: string;
  vehicle_type?: string;
  activity_type?: string;
  owner_id?: string;
  status?: string;
}

class ListingService {
  /**
   * Create a complete listing with location, category details, and images
   */
  async createListing(input: CreateListingInput): Promise<ListingWithDetails> {
    // 1. Create location
    const location = await locationsService.create({
      wilaya: input.location.wilaya,
      city: input.location.city,
      address: input.location.address,
      latitude: input.location.latitude,
      longitude: input.location.longitude,
    });

    // 2. Create post
    const post = await postsService.create({
      owner_id: input.owner_id,
      category: input.category,
      title: input.title,
      description: input.description ?? null,
      price: input.price,
      location_id: location.id,
      status: input.status ?? 'draft',
      availability: input.availability ?? '[]',
    });

    // 3. Create category-specific details
    let stay_details: Stay | undefined;
    let vehicle_details: Vehicle | undefined;
    let activity_details: Activity | undefined;

    if (input.category === 'stay' && input.stay_details) {
      stay_details = await staysService.create({
        post_id: post.id,
        stay_type: input.stay_details.stay_type,
        area: input.stay_details.area ?? null,
        bedrooms: input.stay_details.bedrooms ?? null,
      });
    } else if (input.category === 'vehicle' && input.vehicle_details) {
      vehicle_details = await vehiclesService.create({
        post_id: post.id,
        vehicle_type: input.vehicle_details.vehicle_type,
        model: input.vehicle_details.model ?? null,
        year: input.vehicle_details.year ?? null,
        fuel_type: input.vehicle_details.fuel_type ?? null,
        transmission: input.vehicle_details.transmission ?? false,
        seats: input.vehicle_details.seats ?? null,
        features: (input.vehicle_details.features ?? []) as Json,
      });
    } else if (input.category === 'activity' && input.activity_details) {
      activity_details = await activitiesService.create({
        post_id: post.id,
        activity_type: input.activity_details.activity_type,
        requirements: (input.activity_details.requirements ?? {}) as Json,
      });
    }

    // 4. Upload images
    let images: PostImage[] = [];
    if (input.images && input.images.length > 0) {
      const uploaded = await imageService.uploadImages(post.id, input.images);
      images = uploaded as unknown as PostImage[];
    }

    return {
      ...post,
      location,
      images,
      stay_details,
      vehicle_details,
      activity_details,
    };
  }

  /**
   * Get listing by ID with all related data
   */
  async getListingById(postId: number): Promise<ListingWithDetails> {
    const post = await postsService.getById(postId);
    const location = await locationsService.getById(post.location_id);
    const images = await imageService.getPostImages(postId);

    // Get category-specific details
    let stay_details: Stay | undefined;
    let vehicle_details: Vehicle | undefined;
    let activity_details: Activity | undefined;

    try {
      if (post.category === 'stay') {
        stay_details = await staysService.getById(postId);
      } else if (post.category === 'vehicle') {
        vehicle_details = await vehiclesService.getById(postId);
      } else if (post.category === 'activity') {
        activity_details = await activitiesService.getById(postId);
      }
    } catch {
      // Details might not exist
    }

    // Get average rating
    const { data: ratingData } = await supabase
      .from('reviews')
      .select('rating')
      .eq('post_id', postId);

    const ratings = ratingData ?? [];
    const average_rating =
      ratings.length > 0
        ? ratings.reduce((sum, r) => sum + r.rating, 0) / ratings.length
        : undefined;

    return {
      ...post,
      location,
      images,
      stay_details,
      vehicle_details,
      activity_details,
      average_rating,
      review_count: ratings.length,
    };
  }

  /**
   * Search listings with filters
   */
  async searchListings(
    filters: SearchFilters,
    limit = 20,
    offset = 0
  ): Promise<{ listings: ListingWithDetails[]; total: number }> {
    let query = supabase
      .from('posts')
      .select(
        `
        *,
        location:locations(*),
        images:post_images(*)
      `,
        { count: 'exact' }
      );

    // Only filter by status if explicitly provided
    if (filters.status) {
      query = query.eq('status', filters.status);
    }

    if (filters.category) {
      query = query.eq('category', filters.category);
    }

    if (filters.owner_id) {
      query = query.eq('owner_id', filters.owner_id);
    }

    if (filters.min_price !== undefined) {
      query = query.gte('price', filters.min_price);
    }

    if (filters.max_price !== undefined) {
      query = query.lte('price', filters.max_price);
    }

    if (filters.wilaya) {
      query = query.eq('location.wilaya', filters.wilaya);
    }

    if (filters.city) {
      query = query.eq('location.city', filters.city);
    }

    query = query
      .order('created_at', { ascending: false })
      .range(offset, offset + limit - 1);

    const { data, error, count } = await query;

    if (error) {
      throw new ApiError(500, 'Failed to search listings.', error.message);
    }

    const listings = (data ?? []).map((item: any) => ({
      ...item,
      location: item.location,
      images: item.images ?? [],
    }));

    return { listings, total: count ?? 0 };
  }

  /**
   * Get listings by owner
   */
  async getOwnerListings(ownerId: string): Promise<ListingWithDetails[]> {
    const { listings } = await this.searchListings({ owner_id: ownerId, status: undefined as any }, 100, 0);
    return listings;
  }

  /**
   * Update listing
   */
  async updateListing(
    postId: number,
    updates: Partial<CreateListingInput>
  ): Promise<ListingWithDetails> {
    // Update post fields
    if (updates.title || updates.description || updates.price || updates.status || updates.availability) {
      await postsService.update(postId, {
        ...(updates.title && { title: updates.title }),
        ...(updates.description !== undefined && { description: updates.description }),
        ...(updates.price !== undefined && { price: updates.price }),
        ...(updates.status && { status: updates.status }),
        ...(updates.availability && { availability: updates.availability }),
      });
    }

    // Update location
    const post = await postsService.getById(postId);
    if (updates.location) {
      await locationsService.update(post.location_id, updates.location);
    }

    // Update category-specific details
    if (updates.stay_details) {
      try {
        await staysService.update(postId, updates.stay_details as Partial<Stay>);
      } catch {
        await staysService.create({ post_id: postId, ...updates.stay_details } as any);
      }
    }

    if (updates.vehicle_details) {
      try {
        const vehicleData = {
          ...updates.vehicle_details,
          features: updates.vehicle_details.features as Json | undefined,
        };
        await vehiclesService.update(postId, vehicleData as Partial<Vehicle>);
      } catch {
        await vehiclesService.create({ post_id: postId, ...updates.vehicle_details } as any);
      }
    }

    if (updates.activity_details) {
      try {
        const activityData = {
          ...updates.activity_details,
          requirements: updates.activity_details.requirements as Json | undefined,
        };
        await activitiesService.update(postId, activityData as Partial<Activity>);
      } catch {
        await activitiesService.create({ post_id: postId, ...updates.activity_details } as any);
      }
    }

    // Handle new images - only upload base64 images, skip URLs
    if (updates.images && updates.images.length > 0) {
      const newImages = updates.images.filter(img => 
        !img.startsWith('http://') && !img.startsWith('https://')
      );
      if (newImages.length > 0) {
        await imageService.uploadImages(postId, newImages);
      }
    }

    return this.getListingById(postId);
  }

  /**
   * Delete listing and all related data
   */
  async deleteListing(postId: number): Promise<void> {
    const post = await postsService.getById(postId);

    // Delete images from Cloudinary
    await imageService.deletePostImages(postId);

    // Delete category-specific details (cascade should handle this)
    // Delete reviews (cascade should handle this)
    // Delete bookings (cascade should handle this)

    // Delete post
    await postsService.remove(postId);

    // Delete orphaned location
    await locationsService.remove(post.location_id);
  }

  /**
   * Publish a draft listing
   */
  async publishListing(postId: number): Promise<ListingWithDetails> {
    await postsService.update(postId, { status: 'published' });
    return this.getListingById(postId);
  }

  /**
   * Unpublish a listing
   */
  async unpublishListing(postId: number): Promise<ListingWithDetails> {
    await postsService.update(postId, { status: 'draft' });
    return this.getListingById(postId);
  }
}

export const listingService = new ListingService();
export default listingService;
